runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/expansion/ExpansionService.java

```diff
@@ -69,14 +82,123 @@
+            registrar.knownBuilders().entrySet()) {
+          String urn = entry.getKey();
+          Class<? extends ExternalTransformBuilder<?, ?, ?>> builderClass = entry.getValue();
+          builder.put(
+              urn,
+              spec -> {
+                try {
+                  ExternalTransforms.ExternalConfigurationPayload payload =
+                      ExternalTransforms.ExternalConfigurationPayload.parseFrom(spec.getPayload());
+                  return translate(payload, builderClass);
+                } catch (Exception e) {
+                  throw new RuntimeException(
+                      String.format("Failed to build transform %s from spec %s", urn, spec), e);
+                }
+              });
+        }
+      }
+      return builder.build();
+    }
+
+    private static PTransform translate(
+        ExternalTransforms.ExternalConfigurationPayload payload,
+        Class<? extends ExternalTransformBuilder<?, ?, ?>> builderClass)
+        throws Exception {
+      Preconditions.checkState(
+          ExternalTransformBuilder.class.isAssignableFrom(builderClass),
+          "Provided identifier %s is not an ExternalTransformBuilder.",
+          builderClass.getName());
+
+      Object configObject = initConfiguration(builderClass);
+      populateConfiguration(configObject, payload);
+      return buildTransform(builderClass, configObject);
+    }
+
+    private static Object initConfiguration(
+        Class<? extends ExternalTransformBuilder<?, ?, ?>> builderClass) throws Exception {
+      for (Method method : builderClass.getMethods()) {
+        if (method.getName().equals("buildExternal")) {
+          Preconditions.checkState(
+              method.getParameterCount() == 1,
+              "Build method for ExternalTransformBuilder %s must have exactly one parameter, but had %s parameters.",
+              builderClass.getSimpleName(),
+              method.getParameterCount());
+          Class<?> configurationClass = method.getParameterTypes()[0];
+          if (Object.class.equals(configurationClass)) {
+            continue;
+          }
+          return configurationClass.getDeclaredConstructor().newInstance();
+        }
+      }
+      throw new RuntimeException("Couldn't find build method on ExternalTransformBuilder.");
+    }
+
+    private static void populateConfiguration(
+        Object config, ExternalTransforms.ExternalConfigurationPayload payload) throws Exception {
+      Converter<String, String> camelCaseConverter =
+          CaseFormat.LOWER_UNDERSCORE.converterTo(CaseFormat.LOWER_CAMEL);
+      for (Map.Entry<String, ExternalTransforms.ConfigValue> entry :
+          payload.getConfigurationMap().entrySet()) {
+        String fieldName = camelCaseConverter.convert(entry.getKey());
+        String coderUrn = entry.getValue().getCoderUrn();
+        final Coder coder;
+        if (BeamUrns.getUrn(RunnerApi.StandardCoders.Enum.VARINT).equals(coderUrn)) {
```

Comment:
Shouldn't we be leveraging the coder rehydration code here? Or is that implicit in the TODO below (which should be made more explicit)?